pkg/pillar/cmd/zedrouter/appnumpersist.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Network Instance/underlay network IP Address Management,
+/// persistent App Num management module
+
+package zedrouter
+
+import (
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+//   - Persistent AppNumMap : for all app number assignments
+// For Persistent AppNum Map
+//    - persistentAppNumMapGC()
+//    - persistentAddrMapGet()
+//    - persistentAddrMapAdd()
+//    - persistentAddrMapDelete()
+//    - persistentAddrMapRestoreForNetworkInstance()
+//    - persistentAddrMapDeleteForNetworkInstance()
+//    - persistentAddrMapKey()
+
+// AppNumber garbage collector: for non-existent network instance/
+// app instance, free the persistent appNum map
+func persistentAppNumMapGC(ctx *zedrouterContext) {
+	log.Functionf("persistentAppNumMapGC")
+	freedCount := 0
+	pub := ctx.pubNetworkInstanceAppNumMap
+	items := pub.GetAll()
+	for _, item := range items {
+		addrMap := item.(types.NetworkInstanceAppNumMap)
+		appNetStatus := lookupAppNetworkStatus(ctx, addrMap.AppID)
+		netInstStatus := lookupNetworkInstanceStatus(ctx, addrMap.NetworkID)
+		if netInstStatus == nil || appNetStatus == nil {
+			log.Functionf("persistentAppNumMapGC: freeing %s", addrMap.Key())
+			pub.Unpublish(addrMap.Key())
+			freedCount++
+		}
+	}
+	log.Functionf("persistentAppNumMapGC freed %d", freedCount)
+}
+
+// Persistent Address Map routines
+func persistentAddrMapGet(ctx *zedrouterContext, networkID string,
+	appID string) *types.NetworkInstanceAppNumMap {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	key := persistentAddrMapKey(networkID, appID)
+	if addrMap, _ := pub.Get(key); addrMap != nil {
+		appNumMap := addrMap.(types.NetworkInstanceAppNumMap)
+		return &appNumMap
+	}
+	return nil
+}
+
+func persistentAddrMapAdd(ctx *zedrouterContext,
+	networkID string, appID string, appNumOnUNet int) {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	addrMap := types.NetworkInstanceAppNumMap{
+		NetworkID:    networkID,
+		AppID:        appID,
+		AppNumOnUNet: appNumOnUNet,
+	}
+	pub.Publish(addrMap.Key(), addrMap)
+	return
+}
+
+func persistentAddrMapDelete(ctx *zedrouterContext,
```

Comment:
If we name the above function Allocate it would make sense to name this function Free (instead of Delete).